### PR TITLE
Fixes #150 test issues

### LINKED ISSUE #150: Inconsistent test failures
I am getting test failure when I run the test. Tests seem to fail inconsistently too where one run might give 5 failures and another 3.

Here's an output from one run:

```
17:42 $ npm run unit

> node-xmpp-server@2.1.3 unit /Users/xiaoxin.lu/dev/node-xmpp-server
> mocha --recursive test/ -t 3000



  C2S BOSH server client
    server
      ✓ should listen
    client
could not parse<stream:stream xmlns:stream="http://etherx.jabber.org/streams" xmlns="jabber:client" version="1.0" id="38430efeda84c07ba674ba5797db8b21" from="localhost">
could not parse<stream:stream xmlns:stream="http://etherx.jabber.org/streams" xmlns="jabber:client" version="1.0" id="38430efeda84c07ba674ba5797db8b21" from="localhost">
could not parse<stream:stream xmlns:stream="http://etherx.jabber.org/streams" xmlns="jabber:client" version="1.0" id="38430efeda84c07ba674ba5797db8b21" from="localhost">
      ✓ should connect (80ms)

  C2S TCP server client
    server
      ✓ should listen
    client
      ✓ should connect
      ✓ should disconnect when server shuts down
      1) should connect

  C2S WebSocket server client
    server
      - should listen
    client
      - should connect
      - should disconnect when server shuts down

  SASL
    PLAIN
      ✓ should accept plain authentication
      ✓ should not accept plain authentication
    XOAUTH-2
      ✓ should accept google authentication
    DIGEST MD5
      ✓ should accept digest md5 authentication
      ✓ should not allow to skip digest md5 challenges (506ms)
    ANONYMOUS
      ✓ should accept anonymous authentication

  Stream resource bind
    ✓ Should bind unmodified
    ✓ Should bind modified
    ✓ Should not bind

  Stream register
    ✓ Should register
    ✓ Should not register

  server end
    TCP server
      ✓ disconnects all clients
      ✓ closes the port
    WebSocket server
      ✓ disconnects all clients
      ✓ closes the port
    BOSH server
could not parse<stream:stream xmlns:stream="http://etherx.jabber.org/streams" xmlns="jabber:client" version="1.0" id="7f82de3216271b174acadb16b49e5202" from="localhost">
could not parse<stream:stream xmlns:stream="http://etherx.jabber.org/streams" xmlns="jabber:client" version="1.0" id="7f82de3216271b174acadb16b49e5202" from="localhost">
could not parse<stream:stream xmlns:stream="http://etherx.jabber.org/streams" xmlns="jabber:client" version="1.0" id="7f82de3216271b174acadb16b49e5202" from="localhost">
      ✓ disconnects all clients
      ✓ closes the port

  component server - component
    server
      ✓ should listen
    component
      ✓ should connect
    component not serviced by server
      ✓ should connect
    component supplied invalid credentials
      ✓ should connect
    component that uses wrong namespace
      ✓ should error if wrong namespace

  server stop
    ✓ it closes the port
    ✓ disconnects all clients and shutdown the server
    2) "after each" hook
    BOSHServer


  29 passing (814ms)
  3 pending
  2 failing

  1) C2S TCP server client client should connect:
     Error: read ECONNRESET
      at exports._errnoException (util.js:870:11)
      at TCP.onread (net.js:552:26)

  2) server stop "after each" hook:
     Uncaught Error: read ECONNRESET
      at exports._errnoException (util.js:870:11)
      at TCP.onread (net.js:552:26)
```
I have a PR ready to go to fix this.

### PR DESCRIPTION
**Pull request by hyperlink**
Hopefully this is better. I ran it for awhile using `while npm run unit; do :; done`

### CODE REVIEW

File: test/integration/C2S/TCP.js [hunk 1]
```diff
@@ -38,6 +38,7 @@ describe('C2S TCP server client', function () {
       })
       client.on('error', done)
       client.on('online', function () {
+        client.removeListener('error', done)
```
Comment (sonnyp) [inline test/integration/C2S/TCP.js]: any idea what would make the client emit the error?

File: test/integration/C2S/bind.js [hunk 2]
```diff
@@ -58,46 +60,49 @@ function startClient (cb) {
...
-    startClient(function (error, resource) {
+    client = startClient(function (error, resource) {
       if (error) {
         done(error)
       } else if (resource !== user.jid.resource) {
         done(new Error('Wrong resource: ' + resource))
       } else {
         done()
       }
+      client.end()
```
Comment (sonnyp) [inline test/integration/C2S/bind.js]: client.end should probably called after done

File: test/integration/C2S/bind.js [hunk 3]
```diff
@@ -58,46 +60,49 @@ function startClient (cb) {
...
-    startClient(function (error, resource) {
+    client = startClient(function (error, resource) {
       if (error) {
         done(error)
       } else if (resource !== user.jid.resource + '-' + 'mod') {
         done(new Error('Wrong resource: ' + resource))
       } else {
         done()
       }
+      client.end()
```
Comment (sonnyp) [inline test/integration/C2S/bind.js]: ditto

File: test/integration/C2S/bind.js [hunk 4]
```diff
@@ -58,46 +60,49 @@ function startClient (cb) {
...
   it('Should not bind', function (done) {
     server = startServer('fail')
-    startClient(function (error) {
+    client = startClient(function (error) {
       if (!error) {
         done(new Error('No error'))
       } else {
         done()
       }
+      client.end()
```
Comment (sonnyp) [inline test/integration/C2S/bind.js]: ditto

File: test/integration/C2S/registration.js [hunk 5]
```diff
@@ -60,31 +62,33 @@ function startClient (cb) {
...
   it('Should register', function (done) {
     server = startServer('unmodified')
-    startClient(function (error) {
+    client = startClient(function (error) {
       if (error) {
         done(error)
       } else {
         done()
       }
+      client.end()
```
Comment (sonnyp) [inline test/integration/C2S/registration.js]: ditto

File: test/integration/C2S/registration.js [hunk 6]
```diff
@@ -60,31 +62,33 @@ function startClient (cb) {
...
   it('Should not register', function (done) {
     server = startServer('fail')
-    startClient(function (error) {
+    client = startClient(function (error) {
       if (!error) {
         done(new Error('No error'))
       } else {
         done()
       }
+      client.end()
```
Comment (sonnyp) [inline test/integration/C2S/registration.js]: ditto

File: test/integration/C2S/TCP.js (hunk 1 above)
Comment (hyperlink) [inline test/integration/C2S/TCP.js]: My best guess is when the server shuts down an error is emitted on the client's tcp socket which gets emitted back to the client.

File: test/integration/C2S/bind.js (hunk 2 above)
Comment (hyperlink) [inline test/integration/C2S/bind.js]: Shouldn't `client.end()` be called after each `done`? As in it should be ended independent on the result of the test.

Comment (sonnyp) [inline test/integration/C2S/bind.js]: I meant after done callback

**Comment by sonnyp** [comment]
@hyperlink can you `client.end()` in a afterEach ?

**Comment by sonnyp** [comment]
@hyperlink should I merge?

**Comment by hyperlink** [comment]
Merge would be great @sonnyp thanks!

**Pull request merged**